### PR TITLE
feat(profile-metadata): store profile metadata in warp

### PR DESCRIPTION
**Pull request by ashneverdawn**
### What this PR does 📖

- profile metadata (integrations) used to just be locally stored in the browser. It is now stored in warp. This allows us to access a friend's metadata in subsequent PRs.
- this also makes some changes to allow handling of arbitrary key/value pairs in the profile metadata
- also displays the user's metadata from warp in their profile card (accessible from chat)

### CODE REVIEW

**Review by InfamousVague — APPROVED** [review]

**Comment by luisecm** [comment]
The failing tests on CI are because of the settings profile snapshots tests used to validate changing profile banner/picture are finding a difference between the base snapshot not having this new part for Accounts. I will create a PR in the testing repo to update snapshots and I can merge once this PR is merge too

image

File: src/lib/components/ui/IntegrationDisplay.svelte [hunk 1]
```diff
@@ -2,28 +2,44 @@
...
-    import type { Integration } from "$lib/types"
 
-    export let integration: Integration
-    export let editable: boolean = true
+    export let key: string
+    export let value: string
+    export let editable: boolean = false
 
-    let display = INTEGRATIONS[integration.kind].display
+    function toIntegrationKind(key: string) {
```
[the remark this answers is not in the document]
Comment (ashneverdawn) [reply src/lib/components/ui/IntegrationDisplay.svelte]: done

File: src/lib/components/ui/IntegrationDisplay.svelte [hunk 2]
```diff
@@ -2,28 +2,44 @@
...
+        let integration_kind = Integrations[key as keyof typeof Integrations]
+        if (integration_kind === undefined) {
+            integration_kind = Integrations.Generic
+        }
+        return integration_kind
+    }
+    function display(key: string) {
```
[the remark this answers is not in the document]
Comment (ashneverdawn) [reply src/lib/components/ui/IntegrationDisplay.svelte]: done

File: src/routes/settings/profile/+page.svelte [hunk 3]
```diff
@@ -139,51 +140,49 @@
...
+    let selectedKey: string
+    let selectedKeyEditValue: string
 
     function addIntegration() {
-        if (selectedIntegration.kind && selectedIntegration.location) {
-            let currentIntegrations = get(Store.state.user).integrations || []
-            const updatedIntegrations = [...currentIntegrations, { ...selectedIntegration }]
-            Store.setIntegrations(updatedIntegrations)
-            selectedIntegration = { kind: Integrations.Generic, location: "", meta: "" }
+        if (selectedKey !== "" && selectedKeyEditValue !== "") {
```
[the remark this answers is not in the document]
Comment (ashneverdawn) [reply src/routes/settings/profile/+page.svelte]: done

File: src/routes/settings/profile/+page.svelte [hunk 4]
```diff
@@ -386,63 +385,76 @@
...
+                                        } else {
+                                            selectedKey = selectedKind
+                                        }
+                                    }} />
+                            {/if}
+                            {#if selectedKind === Integrations.Generic}
+                                <Input alt bind:value={selectedKey} disabled={user.integrations.has(selectedKey)} />
+                            {/if}
+                        </div>
+                        <img class="integration-logo" src="/assets/brand/{toIntegrationKind(selectedKey)}.png" alt="Platform Logo" />
```
[the remark this answers is not in the document]
Comment (ashneverdawn) [reply src/routes/settings/profile/+page.svelte]: done

File: src/routes/settings/profile/+page.svelte [hunk 5]
```diff
@@ -386,63 +385,76 @@
...
+                            <Button appearance={Appearance.Error} icon on:click={() => removeIntegration(key)}>
                                 <Icon icon={Shape.XMark} />
                             </Button>
                         </div>
-                    {:else}
+                    {/each}
+                </div>
+
+                {#if $showEditIntegrations}
+                    <Label text={user.integrations.has(selectedKey) ? "Edit Integration" : "Add New"} />
```
[the remark this answers is not in the document]
Comment (ashneverdawn) [reply src/routes/settings/profile/+page.svelte]: done

File: src/routes/settings/profile/+page.svelte [hunk 6]
```diff
@@ -386,63 +385,76 @@
...
+                                <Input alt bind:value={selectedKey} disabled={user.integrations.has(selectedKey)} />
+                            {/if}
+                        </div>
+                        <img class="integration-logo" src="/assets/brand/{toIntegrationKind(selectedKey)}.png" alt="Platform Logo" />
+                        <div class="right">
+                            <Label text="Address" />
+                            <Input alt bind:value={selectedKeyEditValue} />
+                        </div>
+
+                        <Button text={user.integrations.has(selectedKey) ? "Save" : "Add"} on:click={user.integrations.has(selectedKey) ? saveEditedIntegration : addIntegration}>
```
[the remark this answers is not in the document]
Comment (ashneverdawn) [reply src/routes/settings/profile/+page.svelte]: done

File: src/routes/settings/profile/+page.svelte [hunk 7]
```diff
@@ -386,63 +385,76 @@
...
+                            <Icon icon={user.integrations.has(selectedKey) ? Shape.CheckMark : Shape.Plus} />
+                        </Button>
                         <Button
-                            text="Add"
+                            text="Cancel"
```
[the remark this answers is not in the document]
Comment (ashneverdawn) [reply src/routes/settings/profile/+page.svelte]: done

File: src/routes/settings/profile/+page.svelte [hunk 8]
```diff
@@ -139,51 +140,49 @@
...
         }
     }
 
-    function removeIntegration(index: number) {
-        let currentIntegrations = get(Store.state.user).integrations || []
-        currentIntegrations.splice(index, 1)
-        Store.setIntegrations(currentIntegrations)
+    function removeIntegration(key: string) {
+        MultipassStoreInstance.removeMetadata(key)
+        showEditIntegrations.set(false)
```
[the remark this answers is not in the document]
Comment (ashneverdawn) [reply src/routes/settings/profile/+page.svelte:170]: done

File: src/routes/settings/profile/+page.svelte [hunk 9]
```diff
@@ -90,6 +90,7 @@
     })
 
     let user: User = get(Store.state.user)
+    Store.state.user.subscribe(u => (user = u))
```
[the remark this answers is not in the document]
Comment (ashneverdawn) [reply src/routes/settings/profile/+page.svelte]: done

**Comment by ashneverdawn** [comment]
> Hey Hugo! Noticed some weird behavior with the profile pictures while testing
> 
> So when the user is updating their username or status once they hit enter or click save the custom profile picture disappears. When this happens it also doesnt go back to showing the default Uplink image it will just appear blank (so in the screen recording you can see that the picture will appear blank but I tested it again while writing this and after I updated my profile picture the default image did re-appear so not sure what thats about). It also doesnt seem like any strange errors are happening during this action.

I just tried it and I was able to reproduce that on my end. Also, sett banner seems to disappear profile pic and vice versa. Not sure what's causing it. Seems unrelated to what I added in this PR, but I could be missing something.

**Pull request merged**